cmake/multi_image.cmake

```diff
@@ -0,0 +1,164 @@
+  string(TOUPPER ${name} UPNAME)
+
+  if (CONFIG_${UPNAME}_BUILD_STRATEGY_USE_HEX_FILE)
+    assert_exists(CONFIG_${UPNAME}_HEX_FILE)
+    message("Using ${CONFIG_${UPNAME}_HEX_FILE} instead of building ${name}")
+  elseif (CONFIG_${UPNAME}_BUILD_STRATEGY_SKIP_BUILD)
+    message("Skipping building of ${name}")
+  else()
+    # Build normally
+    zephyr_add_external_image_from_source(${name} ${sourcedir})
+  endif()
+endfunction()
+
+function(zephyr_add_external_image_from_source name sourcedir)
+  string(TOUPPER ${name} UPNAME)
+  set(${UPNAME}_CMAKE_BINARY_DIR ${CMAKE_BINARY_DIR}/${name})
+  set(${UPNAME}_PROJECT_BINARY_DIR ${CMAKE_BINARY_DIR}/${name}/zephyr)
+
+  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
+  get_cmake_property(VARIABLES              VARIABLES)
+
+  unset(image_cmake_args)
+  foreach(var_name
+      ${VARIABLES}
+      ${VARIABLES_CACHED}
+      )
+    if("${var_name}" MATCHES "^${name}_(.+)") # TODO: How expensive?
+      list(APPEND image_cmake_args
+        -D${CMAKE_MATCH_1}=${${var_name}}
+        )
+    endif()
+  endforeach()
+
+  file(MAKE_DIRECTORY ${${UPNAME}_CMAKE_BINARY_DIR})
+  message("=== child image ${name} begin ===")
+  image_board_selection(${BOARD} IMAGE_BOARD)
+  image_menuconfig(IMAGE_KCONFIG_TARGET IMAGE_COMMAND_TARGET)
+  execute_process(
+    COMMAND ${CMAKE_COMMAND}
+    # Add other toolchain here,
+    -G${CMAKE_GENERATOR}
+    -DBOARD=${IMAGE_BOARD}
+    ${IMAGE_KCONFIG_TARGET}
+    ${IMAGE_COMMAND_TARGET}
+    -DIS_CHILD_IMAGE=True
+    ${sourcedir}
+    ${image_cmake_args}
+    WORKING_DIRECTORY ${${UPNAME}_CMAKE_BINARY_DIR}
+    RESULT_VARIABLE ret
+    )
+  set(${UPNAME}_DOTCONFIG ${${UPNAME}_PROJECT_BINARY_DIR}/.config)
+  set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${${UPNAME}_DOTCONFIG})
+
+  if(NOT ${ret} EQUAL "0")
+    message(FATAL_ERROR "CMake generation for ${name} failed, aborting. Command: ${ret}")
+  endif()
+
+  message("=== child image ${name} end ===")
+
+  # The variables inhereted from PARENT_SCOPE might be overwritten when additional Kconfig is loaded locally.
+  # Therefore, safe guard the variables from parent scope that are needed.
+  import_kconfig(CONFIG_ ${${UPNAME}_PROJECT_BINARY_DIR}/.config)
+
+  # Clear it, just in case it was defined in parent scope
+  set(spm_veneers_lib)
+  if(CONFIG_${UPNAME}_CONTAINS_SECURE_ENTRIES)
+    set(spm_veneers_lib ${${UPNAME}_CMAKE_BINARY_DIR}/${CONFIG_ARM_ENTRY_VENEERS_LIB_NAME})
+  endif()
+
+  define_property(GLOBAL PROPERTY ${name}_KERNEL_NAME
+                 BRIEF_DOCS "Name (KERNEL_NAME) for the SPM image."
+                 FULL_DOCS "To be used to access e.g. this image's hex file."
+                )
+  set_property(GLOBAL PROPERTY ${name}_KERNEL_NAME ${CONFIG_KERNEL_BIN_NAME})
+
+  include(ExternalProject)
+  ExternalProject_Add(${UPNAME}_subimage
+      SOURCE_DIR ${sourcedir}
+      BINARY_DIR ${${UPNAME}_CMAKE_BINARY_DIR}
+      BUILD_BYPRODUCTS
+       # TODO: Remove this coupling
```

Comment:
Why this comment ?
When using external project, then the best way of handling dependencies is to specify their byproducts, in order to make clear what they produce and to depend on that.
Then use that by product elsewhere in the system in order to ensure correct dependency handling.

Reply:
I mean to say, "Refer to abstract BYPRODUCTS variable instead of spm-specific variable to reduce coupling between multi_image and spm."

Reply:
Got it ;)